src/prefect/utilities/jira_notification.py

```diff
@@ -0,0 +1,123 @@
+"""
+A state-handler that will create and assign a Jira ticket. 
+"""
+
+from typing import TYPE_CHECKING, Union, cast
+from datetime import datetime
+from toolz import curry
+
+import prefect
+
+
+try:
+    from jira import JIRA
+except ImportError:
+    raise ImportError(
+        'Using `jira_notifier` requires Prefect to be installed with the "jira" extra.'
+    )
+
+if TYPE_CHECKING:
+    import prefect.engine.state
+    import prefect.client
+
+TrackedObjectType = Union[prefect.Flow, prefect.Task]
+
+
+def jira_message_formatter(
+    tracked_obj: TrackedObjectType, state: "prefect.engine.state.State"
+) -> str:
+    time = datetime.utcnow()
+    msg = "Message from Prefect: {0} entered a {1} state at {2}".format(
+        tracked_obj.name, type(state).__name__, time
+    )
+    return msg
+
+
+@curry
+def jira_notifier(
+    tracked_obj: TrackedObjectType,
+    old_state: "prefect.engine.state.State",
+    new_state: "prefect.engine.state.State",
+    ignore_states: list = None,
+    only_states: list = None,
+    server_URL: str = None,
+    options: dict = {},
+    assignee: str = "-1",
+) -> "prefect.engine.state.State":
+    """
+    Jira Notifier requires a Jira account and API token.  The API token can be created at: https://id.atlassian.com/manage/api-tokens 
+    The Jira account username ('JIRAUSER'), API token ('JIRATOKEN') should be set as part of a 'JIRASECRETS' object in Prefect Secrets. 
+    An example 'JIRASECRETS' object looks like this:
+    JIRASECRETS = { JIRATOKEN = "XXXXXXXXX", JIRAUSER = "xxxxx@yyy.com", JIRASERVER = "https://???.atlassian.net", JIRAPROJECT = "TEST" }
+    The server URL can be set as part of the 'JIRASECRETS' object ('JIRASERVER') or passed to the jira notifier state handler as the "server_URL" argument.
+    Jira Notifier works as a standalone state handler, or can be called from within a custom
+    state handler.  This function is curried meaning that it can be called multiple times to partially bind any keyword arguments (see example below).
+    Jira Notifier creates a new ticket with the information about the task or flow it is bound to when that task or flow is in a specific state. 
+    (For example it will create a ticket to tell you that the flow you set it on is in a failed state.)  
+    You should use the options dictionary to set the project name and issue type.  
+    You can use the "assignee" argument to assign that ticket to a specific member of your team.
```

Comment:
```suggestion
    Jira Notifier requires a Jira account and API token.  The API token can be created at: https://id.atlassian.com/manage/api-tokens 
    The Jira account username ('JIRAUSER'), API token ('JIRATOKEN') should be set as part of a 'JIRASECRETS' object in Prefect Secrets. 

    An example 'JIRASECRETS' object looks like this:
    ```
    JIRASECRETS = { JIRATOKEN = "XXXXXXXXX", JIRAUSER = "xxxxx@yyy.com", JIRASERVER = "https://???.atlassian.net", JIRAPROJECT = "TEST" }
    ```

    The server URL can be set as part of the 'JIRASECRETS' object ('JIRASERVER') or passed to the jira notifier state handler as the "server_URL" argument.
    Jira Notifier works as a standalone state handler, or can be called from within a custom
    state handler.  This function is curried meaning that it can be called multiple times to partially bind any keyword arguments (see example below).
    Jira Notifier creates a new ticket with the information about the task or flow it is bound to when that task or flow is in a specific state. 
    (For example it will create a ticket to tell you that the flow you set it on is in a failed state.)  
    You should use the options dictionary to set the project name and issue type.  
    You can use the "assignee" argument to assign that ticket to a specific member of your team.
```